### PR TITLE
Fix small reads of multi-part blobs

### PR DESCRIPTION
**Pull request by DaveCTurner**

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java [hunk 1]
```diff
@@ -54,10 +54,13 @@ protected InputStream openInputStream(final long position, final long length) th
...
                     final long currentPart = startPart + slice;
+                    final long startInPart = (currentPart == startPart) ? getRelativePositionInPart(position) : 0L;
+                    final long endInPart
+                        = (currentPart == endPart) ? getRelativePositionInPart(position + length) : getLengthOfPart(currentPart);
                     return blobContainer.readBlob(
                         fileInfo.partName(currentPart),
-                        (currentPart == startPart) ? getRelativePositionInPart(position) : 0L,
-                        (currentPart == endPart) ? getRelativePositionInPart(length) : getLengthOfPart(currentPart)
+                        startInPart,
+                        endInPart - startInPart
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java:63]: Thanks, I was able to reproduce and came with a similar fix.

For better readability I'd suggest to isolate single part case (and be paranoid about the length to read) from the multiple parts case, something like:

```suggestion
        if (fileInfo.numberOfParts() == 1L) {
            return blobContainer.readBlob(fileInfo.partName(0L), position, Math.min(length, fileInfo.partBytes(0) - position));
        }

        final long startPart = getPartNumberForPosition(position);
        final long endPart = getPartNumberForPosition(position + length);

        return new SlicedInputStream(endPart - startPart + 1L) {
            @Override
            protected InputStream openSlice(long slice) throws IOException {
                final long currentPart = startPart + slice;
                final long startInPart = (currentPart == startPart) ? getRelativePositionInPart(position) : 0L;
                final long endInPart
                    = (currentPart == endPart) ? getRelativePositionInPart(position + length) : getLengthOfPart(currentPart);
                return blobContainer.readBlob(fileInfo.partName(currentPart), startInPart, endInPart - startInPart
                );
            }
        };
```

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/TestUtils.java [hunk 2]
```diff
@@ -70,70 +70,100 @@ public static void assertCounter(IndexInputStats.Counter counter, long total, lo
...
-        return new BlobContainer() {
-
+        return new MostlyUnimplementedFakeBlobContainer() {
             @Override
             public InputStream readBlob(String name, long position, long length) throws IOException {
                 if (blobName.equals(name) == false) {
                     throw new FileNotFoundException("Blob not found: " + name);
                 }
-                return Streams.limitStream(new ByteArrayInputStream(blobContent, Math.toIntExact(position), blobContent.length), length);
+                assert position + length <= blobContent.length
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/TestUtils.java:79]: :+1:

**Review by tlrx — APPROVED** [review]
LGTM

Thanks David for catching and fixing this and sorry for this bug. I left a suggestion, feel free to use it or not.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java (hunk 1 above)
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java]: Sure, sounds good, see 558e68a4d0f.

**Pull request merged**